File: Q25.cpp
//Q25.cpp
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode dummy(0);
        dummy.next = head;

        ListNode* prevGroupEnd = &dummy;

        while (true) {
            ListNode* kth = prevGroupEnd;
            // Find the k-th node
            for (int i = 0; i < k && kth; i++) {
                kth = kth->next;
            }
            if (!kth) break; // not enough nodes

            ListNode* groupStart = prevGroupEnd->next;
            ListNode* nextGroupStart = kth->next;

            // Reverse k nodes
            ListNode *prev = nextGroupStart, *curr = groupStart;
            while (curr != nextGroupStart) {
                ListNode* temp = curr->next;
                curr->next = prev;
                prev = curr;
                curr = temp;
            }

            // Connect reversed group with previous part
            prevGroupEnd->next = kth;
            prevGroupEnd = groupStart;
        }

        return dummy.next;
    }
};

File: Q25.py
#Q25.py
class Solution:
    def reverseKGroup(self, head, k):
        dummy = ListNode(0, head)
        prevGroupEnd = dummy

        while True:
            kth = prevGroupEnd
            for _ in range(k):
                kth = kth.next
                if not kth:
                    return dummy.next

            groupStart = prevGroupEnd.next
            nextGroupStart = kth.next

            # Reverse group
            prev, curr = nextGroupStart, groupStart
            while curr != nextGroupStart:
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp

            prevGroupEnd.next = kth
            prevGroupEnd = groupStart
